File: CrackPerformanceGPU/CodigosC/OptimizedBruteForceAttack.c
#include <stdio.h>
#include <string.h>
#include <OpenCL/opencl.h>

#define MAX_PASSWORD_LENGTH 6

// Função para testar uma senha específica
int testPassword(const char *password) {
    // Replace this logic with your own verification mechanism
    const char *correctPassword = "mySecretPassword";

    // Compare the entered password with the correct password
    if (strcmp(password, correctPassword) == 0) {
        return 1; // Correct password
    } else {
        return 0; // Incorrect password
    }
}

// Função para realizar o ataque de força bruta otimizado usando a GPU
void optimizedBruteForceAttackGPU(int maxLength) {
    cl_platform_id platform;
    cl_device_id device;
    cl_context context;
    cl_command_queue queue;
    cl_program program;
    cl_kernel kernel;

    // Inicializar a plataforma OpenCL
    clGetPlatformIDs(1, &platform, NULL);
    clGetDeviceIDs(platform, CL_DEVICE_TYPE_GPU, 1, &device, NULL);
    context = clCreateContext(NULL, 1, &device, NULL, NULL, NULL);
    queue = clCreateCommandQueue(context, device, 0, NULL);

    // Carregar o código do kernel OpenCL
    const char *kernelSource = "__kernel void testPasswords(__global char* passwords) {\n"
                               "    int index = get_global_id(0);\n"
                               "    // Testar a senha correspondente ao índice\n"
                               "    if (testPassword(&passwords[index * (MAX_PASSWORD_LENGTH + 1)])) {\n"
                               "        printf(\"Senha encontrada: %s\\n\", &passwords[index * (MAX_PASSWORD_LENGTH + 1)]);\n"
                               "    }\n"
                               "}\n";
    program = clCreateProgramWithSource(context, 1, &kernelSource, NULL, NULL);
    clBuildProgram(program, 1, &device, NULL, NULL, NULL);

    // Criar o kernel OpenCL
    kernel = clCreateKernel(program, "testPasswords", NULL);

    // Gerar todas as combinações possíveis de senha
    int totalPasswords = 1;
    for (int i = 0; i < maxLength; i++) {
        totalPasswords *= 26;
    }

    char passwords[totalPasswords][MAX_PASSWORD_LENGTH + 1];
    memset(passwords, 0, sizeof(passwords));

    for (int i = 0; i < totalPasswords; i++) {
        // Gerar a senha correspondente ao índice
        int value = i;
        for (int j = 0; j < maxLength; j++) {
            passwords[i][j] = 'a' + (value % 26);
            value /= 26;
        }
    }

    // Criar o buffer de memória para armazenar as senhas na GPU
    cl_mem passwordsBuffer = clCreateBuffer(context, CL_MEM_READ_ONLY, sizeof(passwords), NULL, NULL);
    clEnqueueWriteBuffer(queue, passwordsBuffer, CL_TRUE, 0, sizeof(passwords), passwords, 0, NULL, NULL);

    // Definir os argumentos do kernel
    clSetKernelArg(kernel, 0, sizeof(cl_mem), &passwordsBuffer);

    // Executar o kernel
    size_t globalSize = totalPasswords;
    clEnqueueNDRangeKernel(queue, kernel, 1, NULL, &globalSize, NULL, 0, NULL, NULL);
    clFinish(queue);

    // Liberar recursos
    clReleaseMemObject(passwordsBuffer);
    clReleaseKernel(kernel);
    clReleaseProgram(program);
    clReleaseCommandQueue(queue);
    clReleaseContext(context);
}

int main() {
    int maxLength = MAX_PASSWORD_LENGTH;
    optimizedBruteForceAttackGPU(maxLength);

    return 0;
}

File: CrackPerformanceGPU/Apple/OptimizedBruteForceAttack.swift
/*
Observe que a biblioteca OpenCL não é nativamente suportada em Swift, portanto, é necessário adicionar a biblioteca OpenCL como uma dependência externa para o projeto. Além disso, o código do kernel OpenCL foi mantido em formato de string para ser carregado durante a execução do programa.
*/

import OpenCL

let MAX_PASSWORD_LENGTH = 6

// Função para testar uma senha específica
func testPassword(_ password: String) -> Bool {
    // Replace this logic with your own verification mechanism
    let correctPassword = "mySecretPassword"
    return password == correctPassword
}

// Função para realizar o ataque de força bruta otimizado usando a GPU
func optimizedBruteForceAttackGPU(maxLength: Int) {
    var platform: cl_platform_id?
    var device: cl_device_id?
    var context: cl_context?
    var queue: cl_command_queue?
    var program: cl_program?
    var kernel: cl_kernel?
    
    // Inicializar a plataforma OpenCL
    clGetPlatformIDs(1, &platform, nil)
    clGetDeviceIDs(platform![0], cl_device_type(CL_DEVICE_TYPE_GPU), 1, &device, nil)
    context = clCreateContext(nil, 1, &device!, nil, nil, nil)
    queue = clCreateCommandQueue(context!, device!, 0, nil)
    
    // Carregar o código do kernel OpenCL
    let kernelSource = "__kernel void testPasswords(__global char* passwords) {\n" +
                       "    int index = get_global_id(0);\n" +
                       "    // Testar a senha correspondente ao índice\n" +
                       "    if (testPassword(&passwords[index * (MAX_PASSWORD_LENGTH + 1)])) {\n" +
                       "        printf(\"Senha encontrada: %s\\n\", &passwords[index * (MAX_PASSWORD_LENGTH + 1)]);\n" +
                       "    }\n" +
                       "}\n"
    let kernelCString = kernelSource.cString(using: .utf8)!
    program = clCreateProgramWithSource(context!, 1, &kernelCString, nil, nil)
    clBuildProgram(program!, 1, &device!, nil, nil, nil)
    
    // Criar o kernel OpenCL
    kernel = clCreateKernel(program!, "testPasswords", nil)
    
    // Gerar todas as combinações possíveis de senha
    var totalPasswords = 1
    for _ in 0..<maxLength {
        totalPasswords *= 26
    }
    
    var passwords = Array(repeating: Array(repeating: CChar(0), count: MAX_PASSWORD_LENGTH + 1), count: totalPasswords)
    
    for i in 0..<totalPasswords {
        // Gerar a senha correspondente ao índice
        var value = i
        for j in 0..<maxLength {
            passwords[i][j] = CChar(97 + (value % 26))
            value /= 26
        }
    }
    
    // Criar o buffer de memória para armazenar as senhas na GPU
    let passwordsBuffer = clCreateBuffer(context!, cl_mem_flags(CL_MEM_READ_ONLY), MemoryLayout<CChar>.stride * (MAX_PASSWORD_LENGTH + 1) * totalPasswords, nil, nil)
    clEnqueueWriteBuffer(queue!, passwordsBuffer, cl_bool(CL_TRUE), 0, MemoryLayout<CChar>.stride * (MAX_PASSWORD_LENGTH + 1) * totalPasswords, passwords, 0, nil, nil)
    
    // Definir os argumentos do kernel
    clSetKernelArg(kernel!, 0, MemoryLayout<cl_mem>.stride, &passwordsBuffer)
    
    // Executar o kernel
    var globalSize = size_t(totalPasswords)
    clEnqueueNDRangeKernel(queue!, kernel!, 1, nil, &globalSize, nil, 0, nil, nil)
    clFinish(queue!)
    
    // Liberar recursos
    clReleaseMemObject(passwordsBuffer)
    clReleaseKernel(kernel!)
    clReleaseProgram(program!)
    clReleaseCommandQueue(queue!)
    clReleaseContext(context!)
}

let maxLength = MAX_PASSWORD_LENGTH
optimizedBruteForceAttackGPU(maxLength: maxLength)

